src/devices/sound/discrete.cpp

```diff
@@ -91,61 +91,70 @@ DEFINE_DEVICE_TYPE(DISCRETE, discrete_sound_device, "discrete", "Discrete Sound"
 
 struct output_buffer
 {
+	output_buffer() : source(nullptr), ptr(nullptr)
+	{
+	}
+
+	// required for use in vector (std::atomic<T> has deleted copy constructor)
+	output_buffer(output_buffer &&that) : node_buf(std::move(that.node_buf)), source(that.source), ptr(that.ptr.load(std::memory_order_relaxed)), node_num(that.node_num)
+	{
+	}
+
 	std::unique_ptr<double []>  node_buf;
-	const double                *source;
-	volatile double             *ptr;
+	const double *              source;
+	std::atomic<double *>       ptr;
 	int                         node_num;
 };
 
 struct input_buffer
 {
-	volatile const double       *ptr;               /* pointer into linked_outbuf.nodebuf */
-	output_buffer *             linked_outbuf;      /* what output are we connected to ? */
-	double                      buffer;             /* input[] will point here */
+	const double *              ptr;                // pointer into linked_outbuf.nodebuf
+	output_buffer *             linked_outbuf;      // what output are we connected to?
+	double                      buffer;             // input[] will point here
```

Comment:
There’s no point having `volatile` on `ptr` here, either.  You just need to ensure that there’s an acquire barrier between determining the number of samples available and reading the samples.